### PR TITLE
Async support for vector index creation

### PR DESCRIPTION
**Pull request by emptycrown**

### CODE REVIEW

File: tests/indices/vector_store/test_base.py [hunk 1]
```diff
@@ -272,7 +299,7 @@ def test_simple_insert(
     documents: List[Document],
     struct_kwargs: Dict,
 ) -> None:
-    """Test build GPTFaissIndex."""
+    """Test insert GPTSimpleVectorIndex."""
```
Comment (jerryjliu) [inline tests/indices/vector_store/test_base.py:302]: oops

**Review by jerryjliu — COMMENTED** [review]
I think this works. It is a decent amount of redundant code which I'm wary of. The only other option I was going to suggest is if you just added async to build_index_from_document and not the embeddings part, but that would still introduce blocking embedding API calls, so speed gain probably wouldn't be as good.

I'm ok with implementing the rest of the vector stores like this! If you happened to have a notebook, similar to what @Disiok showed in https://github.com/jerryjliu/gpt_index/tree/main/examples/async (maybe you could append to one of these?), that would be great for demo purposes 🙏

**Comment by emptycrown** [comment]
> I think this works. It is a decent amount of redundant code which I'm wary of. The only other option I was going to suggest is if you just added async to build_index_from_document and not the embeddings part, but that would still introduce blocking embedding API calls, so speed gain probably wouldn't be as good.

Thanks for the thoughts. Yeah re the redundant code -- is there another way? Some other codebases I saw also had similar setups where the functions are essentially copy-pasted. JS makes this way easier but I didn't find a way around it.

Also, why would the embedding API calls be blocking? I switched to the OpenAI async API and then converted everything to async all the way.

File: gpt_index/embeddings/base.py [hunk 2]
```diff
@@ -76,6 +82,16 @@ def _get_text_embeddings(self, texts: List[str]) -> List[List[float]]:
...
         return result
 
+    async def _aget_text_embeddings(self, texts: List[str]) -> List[List[float]]:
+        """Asynchronously get text embeddings.
+
+        By default, this is a wrapper around _aget_text_embedding.
+        Can be overriden for batch queries.
+
+        """
+        result = [await self._aget_text_embedding(text) for text in texts]
```
Comment (TmLev) [inline gpt_index/embeddings/base.py:92]: This looks sequential. Shouldn't this be more like
```suggestion
        #               ▼ Depends on `asyncio` runtime.
        result = await asyncio.gather(*[
            self._aget_text_embedding(text) 
            for text in texts
        ])
```
?

File: gpt_index/embeddings/base.py [hunk 3]
```diff
@@ -117,6 +133,32 @@ def get_queued_text_embeddings(self) -> Tuple[List[str], List[List[float]]]:
...
+        result_embeddings: List[List[float]] = []
+        for idx, (text_id, text) in enumerate(text_queue):
+            cur_batch.append((text_id, text))
+            text_tokens_count = len(self._tokenizer(text))
+            self._total_tokens_used += text_tokens_count
+            if idx == len(text_queue) - 1 or len(cur_batch) == self._embed_batch_size:
+                # flush
+                cur_batch_ids = [text_id for text_id, _ in cur_batch]
+                cur_batch_texts = [text for _, text in cur_batch]
+                embeddings = await self._aget_text_embeddings(cur_batch_texts)
```
Comment (TmLev) [inline gpt_index/embeddings/base.py:156]: Here, too.

File: gpt_index/embeddings/base.py (hunk 2 above)
Comment (emptycrown) [reply gpt_index/embeddings/base.py:92]: Yeah you're right, though this gets overridden anyways, but agreed I'll make the change

**Comment by jerryjliu** [comment]
> Also 2 open questions:
> 
> * How to get asyncio to work with pytest? Just mocking run_async_tasks doesn't work here

Tbh I don't really know, closest I found is this https://tonybaloney.github.io/posts/async-test-patterns-for-pytest-and-unittest.html

> * Not sure what we want to do with the aget_text_embedding class function in BaseEmbedding since only OpenAI has the async api

we can always just call the synchronous get_text_embedding function as a default right?

**Comment by jerryjliu** [comment]
> Also, why would the embedding API calls be blocking? I switched to the OpenAI async API and then converted everything to async all the way up.

No i meant in the hypothetical scenario i suggested where you async the outer function but the inside is still a synchronous call to OpenAI's API (which i said would be slow)

**Comment by emptycrown** [comment]
> No i meant in the hypothetical scenario i suggested where you async the outer function but the inside is still a synchronous call to OpenAI's API (which i said would be slow)

I see -- yeah the gains are mainly from async-ing the Embedding API calls. Anyways, the current version should all be working! **Added a ipynb demo'ing some big gains.** 

Also, I was able to fix the tests by rewriting `run_async_tasks` to what I think is the correct way? cc @Disiok

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**